File: frontend/src/api/timesheets.ts
import client from "./client"
import type {
  TimesheetListResponse,
  TimesheetEntry,
  TimesheetEntryCreate,
  TimesheetEntryUpdate,
  TimesheetEditHistory,
  HrmsSyncLog,
  HrmsSyncLogsResponse,
} from "@/types/timesheet"

export async function getTimesheets(params: {
  employee_id?: string
  project_id?: string
  period?: string
  status?: string
  page?: number
  page_size?: number
}): Promise<TimesheetListResponse> {
  const res = await client.get<TimesheetListResponse>("/timesheets", { params })
  return res.data
}

export async function createTimesheetEntry(
  data: TimesheetEntryCreate
): Promise<TimesheetEntry> {
  const res = await client.post<TimesheetEntry>("/timesheets", data)
  return res.data
}

export async function updateTimesheetEntry(
  entryId: string,
  data: TimesheetEntryUpdate
): Promise<TimesheetEntry> {
  const res = await client.put<TimesheetEntry>(`/timesheets/${entryId}`, data)
  return res.data
}

export async function deleteTimesheetEntry(entryId: string): Promise<void> {
  await client.delete(`/timesheets/${entryId}`)
}

export async function submitTimesheetEntries(
  entryIds: string[]
): Promise<{ submitted_count: number }> {
  const res = await client.post<{ submitted_count: number }>("/timesheets/submit", {
    entry_ids: entryIds,
  })
  return res.data
}

export async function approveRejectEntries(params: {
  entry_ids: string[]
  action: "approve" | "reject"
  rejection_reason?: string
}): Promise<{ processed_count: number }> {
  const res = await client.post<{ processed_count: number }>("/timesheets/approve", params)
  return res.data
}

export async function getEntryHistory(
  entryId: string
): Promise<TimesheetEditHistory[]> {
  const res = await client.get<TimesheetEditHistory[]>(`/timesheets/${entryId}/history`)
  return res.data
}

export async function checkPeriodLock(
  period: string
): Promise<{ period: string; is_locked: boolean }> {
  const res = await client.get<{ period: string; is_locked: boolean }>(
    "/timesheets/period-lock",
    { params: { period } }
  )
  return res.data
}

export async function togglePeriodLock(
  period: string,
  lock: boolean
): Promise<{ period: string; is_locked: boolean }> {
  const res = await client.post<{ period: string; is_locked: boolean }>(
    "/timesheets/period-lock",
    { period, lock }
  )
  return res.data
}

export async function triggerHrmsSync(period: string): Promise<HrmsSyncLog> {
  const res = await client.post<HrmsSyncLog>("/hrms-sync/trigger", { period })
  return res.data
}

export async function getHrmsSyncLogs(params?: {
  page?: number
  page_size?: number
}): Promise<HrmsSyncLogsResponse> {
  const res = await client.get<HrmsSyncLogsResponse>("/hrms-sync/logs", { params })
  return res.data
}

export async function approveTimesheetEntries(
  entryIds: string[]
): Promise<{ processed_count: number }> {
  return approveRejectEntries({ entry_ids: entryIds, action: "approve" })
}

export async function rejectTimesheetEntries(
  entryIds: string[],
  reason: string
): Promise<{ processed_count: number }> {
  return approveRejectEntries({
    entry_ids: entryIds,
    action: "reject",
    rejection_reason: reason,
  })
}

export async function getPeriodLockStatus(
  _locationId: string,
  period: string
): Promise<{ locked: boolean; period: string }> {
  const data = await checkPeriodLock(period)
  return { locked: data.is_locked, period: data.period }
}

export async function getProjects(
  _locationId: string
): Promise<{ id: string; name: string }[]> {
  const res = await client.get<{ id: string; name: string }[]>("/org/projects")
  return res.data
}

export async function getEmployees(
  _locationId: string
): Promise<{ id: string; name: string }[]> {
  const res = await client.get<{ id: string; name: string }[]>("/org/employees")
  return res.data
}

File: backend/app/routers/timesheets.py
from fastapi import APIRouter, Depends, HTTPException, Query

from app.middleware.auth_middleware import CurrentUser, get_current_user
from app.schemas.timesheet import (
    TimesheetEntryCreate,
    TimesheetEntryUpdate,
    TimesheetSubmitRequest,
    TimesheetApprovalRequest,
    PeriodLockRequest,
)
from app.services import timesheet_service

router = APIRouter(prefix="/timesheets", tags=["Timesheets"])


@router.post("/")
async def create_entry(
    body: TimesheetEntryCreate,
    user: CurrentUser = Depends(get_current_user),
):
    try:
        return await timesheet_service.create_entry(body.model_dump(), user)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.put("/{entry_id}")
async def update_entry(
    entry_id: str,
    body: TimesheetEntryUpdate,
    user: CurrentUser = Depends(get_current_user),
):
    try:
        return await timesheet_service.update_entry(
            entry_id, body.model_dump(exclude_unset=True), user
        )
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.delete("/{entry_id}")
async def delete_entry(
    entry_id: str,
    user: CurrentUser = Depends(get_current_user),
):
    try:
        return await timesheet_service.delete_entry(entry_id, user)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.get("/")
async def list_entries(
    employee_id: str = Query(None),
    project_id: str = Query(None),
    period: str = Query(None),
    status: str = Query(None),
    page: int = Query(1, ge=1),
    page_size: int = Query(50, ge=1, le=200),
    user: CurrentUser = Depends(get_current_user),
):
    return await timesheet_service.list_entries(
        employee_id=employee_id,
        project_id=project_id,
        period=period,
        status=status,
        branch_location_id=user.branch_location_id,
        page=page,
        page_size=page_size,
    )


@router.post("/submit")
async def submit_entries(
    body: TimesheetSubmitRequest,
    user: CurrentUser = Depends(get_current_user),
):
    return await timesheet_service.submit_entries(body.entry_ids, user)


@router.post("/approve")
async def approve_reject_entries(
    body: TimesheetApprovalRequest,
    user: CurrentUser = Depends(get_current_user),
):
    try:
        return await timesheet_service.approve_reject_entries(
            entry_ids=body.entry_ids,
            action=body.action,
            rejection_reason=body.rejection_reason,
            user=user,
        )
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.get("/heatmap")
async def get_workload_heatmap(
    period: str = Query(..., pattern=r"^\d{4}-\d{2}$"),
    user: CurrentUser = Depends(get_current_user),
):
    return await timesheet_service.get_workload_heatmap(
        period=period,
        branch_location_id=user.branch_location_id,
    )


@router.get("/period-lock")
async def check_period_lock(
    period: str = Query(...),
    user: CurrentUser = Depends(get_current_user),
):
    return await timesheet_service.check_period_lock(period, user.branch_location_id)


@router.get("/{entry_id}/history")
async def get_entry_history(
    entry_id: str,
    user: CurrentUser = Depends(get_current_user),
):
    return await timesheet_service.get_entry_history(entry_id)


@router.post("/period-lock")
async def toggle_period_lock(
    body: PeriodLockRequest,
    user: CurrentUser = Depends(get_current_user),
):
    return await timesheet_service.toggle_period_lock(
        period=body.period,
        branch_location_id=user.branch_location_id,
        lock=body.lock,
        user=user,
    )
